### PR TITLE
Allow each team control their pod timeout

### PR DESCRIPTION
**Pull request by a-berahman**
Reference to issue https://github.com/kubernetes/test-infra/issues/26614

### CODE REVIEW

File: prow/apis/prowjobs/v1/types.go [hunk 1]
```diff
@@ -487,6 +487,16 @@ type DecorationConfig struct {
 	// set the same as this request. Could be overridden by memory request
 	// defined explicitly on prowjob.
 	DefaultMemoryRequest *resource.Quantity `json:"default_memory_request,omitempty"`
+
+	// PodPendingTimeout is after how long the controller will perform a garbage
+	// collection on pending pods. Defaults to 10 minutes.
```
Comment (listx) [inline prow/apis/prowjobs/v1/types.go:492]: 1. Are these defaults what we already have today? (If they're different, can you explain why?)
2. Can you add integration tests, if possible, for these new fields?

File: prow/apis/prowjobs/v1/types.go [hunk 2]
```diff
@@ -697,6 +707,17 @@ func (d *DecorationConfig) ApplyDefault(def *DecorationConfig) *DecorationConfig
...
+		merged.PodPendingTimeout = def.PodPendingTimeout
+	}
+
+	if merged.PodRunningTimeout == nil {
+		merged.PodRunningTimeout = def.PodRunningTimeout
+	}
+
+	if merged.PodUnscheduledTimeout == nil {
+		merged.PodUnscheduledTimeout = def.PodUnscheduledTimeout
+	}
```
Comment (petr-muller) [inline prow/apis/prowjobs/v1/types.go:720]: Maybe I'm missing something, but how are the values consumed? What actually enforces these timeouts?

**Comment by cjwagner** [comment]
In addition to the points raised by Linus and Petr, we also need to fallback to using the globally configured values if the decoration config values are unspecified (as will be the case for all existing jobs).

File: prow/apis/prowjobs/v1/types.go (hunk 1 above)
Comment (a-berahman) [reply prow/apis/prowjobs/v1/types.go:527]: 1. according to this part, https://github.com/kubernetes/test-infra/blob/master/prow/config/prow-config-documented.yaml#L1087, we have three variables I considered them.
2. sure, I will do it shortly

Comment (a-berahman) [reply prow/apis/prowjobs/v1/types.go:527]: > 1. Are these defaults what we already have today? (If they're different, can you explain why?)
> 2. Can you add integration tests, if possible, for these new fields?

1. 1according to this part, https://github.com/kubernetes/test-infra/blob/master/prow/config/prow-config-documented.yaml#L1087, we have three variables I considered them.
2. sure, I will do it shortly

File: prow/apis/prowjobs/v1/types.go (hunk 2 above)
Comment (a-berahman) [reply prow/apis/prowjobs/v1/types.go:755]: > Maybe I'm missing something, but how are the values consumed? What actually enforces these timeouts?

Prow allows setting DecorationConfig on both prowjob yaml and prow config yaml, and the values of these two are merged at runtime by ApplyDefault function https://github.com/kubernetes/test-infra/blob/959d4500863c43efa5cb45b49ba6d14c62441103/prow/apis/prowjobs/v1/types.go#L615 
@chaodaiG please correct me if I'm wrong

**Comment by a-berahman** [comment]
@cjwagner IIRC I answered your question in my reply to @petr-muller if it's not the case that you mentioned, could you please elaborate more then?

**Comment by cjwagner** [comment]
> > In addition to the points raised by Linus and Petr, we also need to fallback to using the globally configured values if the decoration config values are unspecified (as will be the case for all existing jobs).
> 
> @cjwagner IIRC I answered your question in my reply to @petr-muller if it's not the case that you mentioned, could you please elaborate more then? [#28380 (comment)](https://github.com/kubernetes/test-infra/pull/28380#discussion_r1068978249)

When I say the globally configured values I'm not talking about the decorationconfig defaults in config.yaml I'm talking about the way that pod timeouts were previously globally configured via `plank`'s config. https://github.com/kubernetes/test-infra/blob/21c175ddf29e082200945775a2d8600e30403369/prow/config/config.go#L618-L626

We need to fallback to using these globally configured values if the decorationconfig does not specify the value.

As Petr pointed out, the fields you added are never actually used. The `ApplyDefaults` function you referenced merges the configuration, it does not use the configuration to affect timeouts. As is, this PR allows us to configure timeouts at the job level, but those timeouts would not be respected or have any effect. See how the timeouts in the global `plank` config are used by `prow-controller-manager` to see what I mean.

**Comment by a-berahman** [comment]
> > > In addition to the points raised by Linus and Petr, we also need to fallback to using the globally configured values if the decoration config values are unspecified (as will be the case for all existing jobs).
> > 
> > 
> > @cjwagner IIRC I answered your question in my reply to @petr-muller if it's not the case that you mentioned, could you please elaborate more then? [#28380 (comment)](https://github.com/kubernetes/test-infra/pull/28380#discussion_r1068978249)
> 
> When I say the globally configured values I'm not talking about the decorationconfig defaults in config.yaml I'm talking about the way that pod timeouts were previously globally configured via `plank`'s config.
> 
> https://github.com/kubernetes/test-infra/blob/21c175ddf29e082200945775a2d8600e30403369/prow/config/config.go#L618-L626
> 
> We need to fallback to using these globally configured values if the decorationconfig does not specify the value.
> 
> As Petr pointed out, the fields you added are never actually used. The `ApplyDefaults` function you referenced merges the configuration, it does not use the configuration to affect timeouts. As is, this PR allows us to configure timeouts at the job level, but those timeouts would not be respected or have any effect. See how the timeouts in the global `plank` config are used by `prow-controller-manager` to see what I mean.

Thanks for clarification!
Let me check the mentioned point, I will make changes ASAP

**Pull request closed without merging**